### ISSUE TITLE
Unable to convert client data to unicode

### ISSUE THREAD

**Issue by jaraco** [cannatag/ldap3#511]
In ldap3 2.4.0, the code that queries thus would work:

```
from __future__ import unicode_literals
from ldap3.utils.conv import escape_filter_chars as eb
# (...) set up connection and search root
username = 'gutiérrez'
filter = '(sAMAccountName=' + eb(username) + ')'
conn.search(search_root, filter)
```

Starting with ldap3 2.4.1 on Python 2.7 only, this code fails with:

```
LDAPMaximumRetriesError: LDAPMaximumRetriesError: restartable connection failed to send/Exception history:/    0 2018-03-21 17:28:38.251350: <type 'exceptions.UnicodeError'> - Unable to convert client data to unicode: 'guti\xc3\xa9rrez'/Maximum number of retries reached: 3
```

This issue is possibly related to #482. I'll test against master.

**Comment by jaraco** [comment]
The issue persists even with master. What's the recommended way to construct a filter with unicode values?

**Comment by jaraco** [comment]
The way I read it, the eb(username) won't have any effect on that code. so the filter could be simply `'(sAMAccountName=gutiérrez)'`.

**Comment by cannatag** [comment]
Yes, the filter should be a unicode string. The escape_filter_chars() is used to convert some invalid characters to their ldap representation (for example ( ) \ * and the NUL value). 

Do you still have errors? What python and pyasn1 versions are you using?

**Comment by jaraco** [comment]
Yes, we still have errors, but when I tried to distill the issue to a simple script running ldap3, I couldn't replicate the failure. After bisecting different aspects of our test suite, I eventually distilled the issue to being triggered by the use of pytest. Consider this distilled example:

```
$ cd /tmp/test-issue-511
$ python -m virtualenv --python python2.7 .env
Running virtualenv with interpreter /usr/local/opt/python@2/bin/python2.7
New python executable in /private/tmp/test-issue-511/.env/bin/python2.7
Also creating executable in /private/tmp/test-issue-511/.env/bin/python
Installing setuptools, pip, wheel...done.
$ .env/bin/pip install pytest ldap3
Collecting pytest
  Using cached pytest-3.5.0-py2.py3-none-any.whl
Collecting ldap3
  Using cached ldap3-2.4.1-py2.py3-none-any.whl
Collecting py>=1.5.0 (from pytest)
  Using cached py-1.5.3-py2.py3-none-any.whl
Requirement already satisfied: setuptools in ./.env/lib/python2.7/site-packages (from pytest)
Collecting more-itertools>=4.0.0 (from pytest)
  Using cached more_itertools-4.1.0-py2-none-any.whl
Collecting pluggy<0.7,>=0.5 (from pytest)
Collecting funcsigs; python_version < "3.0" (from pytest)
  Using cached funcsigs-1.0.2-py2.py3-none-any.whl
Collecting attrs>=17.4.0 (from pytest)
  Using cached attrs-17.4.0-py2.py3-none-any.whl
Collecting six>=1.10.0 (from pytest)
  Using cached six-1.11.0-py2.py3-none-any.whl
Collecting pyasn1>=0.1.8 (from ldap3)
  Using cached pyasn1-0.4.2-py2.py3-none-any.whl
Installing collected packages: py, six, more-itertools, pluggy, funcsigs, attrs, pytest, pyasn1, ldap3
Successfully installed attrs-17.4.0 funcsigs-1.0.2 ldap3-2.4.1 more-itertools-4.1.0 pluggy-0.6.0 py-1.5.3 pyasn1-0.4.2 pytest-3.5.0 six-1.11.0
$ cat unicode-test.py
# coding: utf-8

from __future__ import unicode_literals

import os
import ldap3
from ldap3.utils.conv import escape_filter_chars as eb

import six


def test_search():
    password = os.environ['LDAP_PASSWORD']
    username = os.environ['LDAP_USERNAME']
    root = os.environ['LDAP_ROOT']
    hostname = 'ldap://ldap.yougov.local'
    server = ldap3.Server(hostname)
    conn = ldap3.Connection(
        server, user=username,
        password=six.text_type(password))
    conn.open()
    assert conn.bind()
    name = 'é'
    filter = '(sAMAccountName=' + eb(name) + ')'
    conn.search(root, filter)


__name__ == '__main__' and test_search()
$ .env/bin/python unicode-test.py
$ .env/bin/pytest unicode-test.py
========================================================================= test session starts =========================================================================
platform darwin -- Python 2.7.14, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /private/tmp/test-issue-511, inifile:
collected 1 item

unicode-test.py F                                                                                                                                               [100%]

============================================================================== FAILURES ===============================================================================
_____________________________________________________________________________ test_search _____________________________________________________________________________

    def test_search():
        password = os.environ['LDAP_PASSWORD']
        username = os.environ['LDAP_USERNAME']
        root = os.environ['LDAP_ROOT']
        hostname = 'ldap://ldap.yougov.local'
        server = ldap3.Server(hostname)
        conn = ldap3.Connection(
            server, user=username,
            password=six.text_type(password))
        conn.open()
        assert conn.bind()
        name = 'é'
        filter = '(sAMAccountName=' + eb(name) + ')'
>       conn.search(root, filter)

unicode-test.py:25:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.env/lib/python2.7/site-packages/ldap3/core/connection.py:775: in search
    response = self.post_send_search(self.send('searchRequest', request, controls))
.env/lib/python2.7/site-packages/ldap3/strategy/base.py:303: in send
    self.connection.request = BaseStrategy.decode_request(message_type, request, controls)
.env/lib/python2.7/site-packages/ldap3/strategy/base.py:635: in decode_request
    result = search_request_to_dict(component)
.env/lib/python2.7/site-packages/ldap3/operation/search.py:516: in search_request_to_dict
    'filter': filter_to_string(request['filter']),
.env/lib/python2.7/site-packages/ldap3/operation/search.py:474: in filter_to_string
    ava = ava_to_dict(filter_object['equalityMatch'])
.env/lib/python2.7/site-packages/ldap3/protocol/convert.py:87: in ava_to_dict
    return {'attribute': str(ava['attributeDesc']), 'value': escape_filter_chars(str(ava['assertionValue']))}
.env/lib/python2.7/site-packages/ldap3/utils/conv.py:91: in escape_filter_chars
    text = to_unicode(text, encoding)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

obj = '\xc3\xa9', encoding = 'ascii', from_server = False

    def to_unicode(obj, encoding=None, from_server=False):
        """Try to convert bytes (and str in python2) to unicode.
         Return object unmodified if python3 string, else raise an exception
        """
        conf_default_client_encoding = get_config_parameter('DEFAULT_CLIENT_ENCODING')
        conf_default_server_encoding = get_config_parameter('DEFAULT_SERVER_ENCODING')
        conf_additional_server_encodings = get_config_parameter('ADDITIONAL_SERVER_ENCODINGS')
        if isinstance(obj, NUMERIC_TYPES):
            obj = str(obj)

        if isinstance(obj, (bytes, bytearray)):
            if from_server:  # data from server
                if encoding is None:
                    encoding = conf_default_server_encoding
                try:
                    return obj.decode(encoding)
                except UnicodeDecodeError:
                    for encoding in conf_additional_server_encodings:  # AD could have DN not encoded in utf-8 (even if this is not allowed by RFC4510)
                        try:
                            return obj.decode(encoding)
                        except UnicodeDecodeError:
                            pass
                    raise UnicodeError("Unable to convert server data to unicode: %r" % obj)
            else:  # data from client
                if encoding is None:
                    encoding = conf_default_client_encoding
                try:
                    return obj.decode(encoding)
                except UnicodeDecodeError:
>                   raise UnicodeError("Unable to convert client data to unicode: %r" % obj)
E                   UnicodeError: Unable to convert client data to unicode: '\xc3\xa9'

.env/lib/python2.7/site-packages/ldap3/utils/conv.py:64: UnicodeError
====================================================================== 1 failed in 3.29 seconds =======================================================================
```

As you can see, the example script runs fine in the environment when invoked naturally, but fails when run under pytest. My suspicion now is that we've stumbled on a bug in pytest. I'll continue to investigate here.

**Comment by jaraco** [comment]
That value, `'é'`, goes into the conn.search as unicode: `u'(sAMAccountName=\xe9)'`, but somehow gets encoded to utf-8 before it reaches the `to_unicode` call. I'll keep searching to see if I can determine when/how that happens.

**Comment by jaraco** [comment]
At the point where the error occurs, `ldap.get_config_parameter('DEFAULT_ENCODING')` returns the two alternate encodings (ascii under pytest).

**Comment by jaraco** [comment]
Okay. I've figured it out. The issue is triggered by the output capturing. When pytest runs, it replaces `sys.stdin` with a wrapper in order to capture standard input, which affects the logic in `ldap3.utils.config`. If I run `pytest` with `-s` (don't capture input/output), the tests pass once again.

The [offending logic is here](https://github.com/cannatag/ldap3/blob/12106a9467162767c5772544957bd822eef37a02/ldap3/utils/config.py#L94-L99). When pytest is capturing the stdin, the first branch is False, so the default encoding gets set to `sys.getdefaultencoding()`, which is ASCII.

So even though in the usage, I'm passing values as unicode, then in `parse_filter`, they're getting encoded as UTF-8 and then later attempted to be decoded using `sys.getdefaultencoding()`, and all as part of the client handling (not encoding for the server yet).

**Comment by jaraco** [comment]
That also explains why the tests pass on Python 3 where `sys.getdefaultencoding()` is `utf-8`.

@cannatag I imagine I could work around the issue by patching that value in pytest. Better would be for ldap3 to provide its own pytest plugin to patch itself, but even better would be to have the underlying library to avoid encoding/decoding values in the filter creation, though I realize that may be easier said than done. Do you have any questions on the description of the issue above? Do you know why the 2.4.1 release introduced this failure? What would you recommend going forward?

**Comment by cannatag** [comment]
Hi, have you tried to set auto_encode and auto_escape to False while defining the Connection object?

**Comment by jaraco** [comment]
I have not. I'll have to read up on those options.

**Comment by jaraco** [comment]
The same error occurs even with both values set to False.

**Comment by jaraco** [comment]
Git bisect implicates aacdf9ec432d2705148b5d2ae9655fe77ab50164.

**Comment by cannatag** [comment]
Hi, I tried to replicate with pytest and python2.7 in a virtualenv in windows but I don't get the error. I see you're on OSX. Will try on a Mac tomorrow.

**Comment by jaraco** [comment]
That's strange. We've encountered the issue on Linux as well as macOS, so I would expect it to behave the same on Windows.

In fact, I was able to replicate the issue on Windows all the same using the same script as above.

Can you confirm that in your Python 2.7 environment,

```
>>> sys.getdefaultencoding()
'ascii'
```

Here's a more distilled test that doesn't touch the server that reveals the underlying issue:

```
$ cat test-encoding.py
import sys
import ldap3.utils.config


def test_encodings():
    assert sys.getdefaultencoding() == 'ascii'
    assert ldap3.utils.config._DEFAULT_CLIENT_ENCODING == 'UTF-8'


__name__ == '__main__' and test_encodings()
$ python2.7 -m rwt -q pytest ldap3 -- test-encoding.py
$ python2.7 -m rwt -q pytest ldap3 -- -m pytest test-encoding.py
========================================================================== test session starts ==========================================================================
platform darwin -- Python 2.7.14, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /Users/jaraco/Dropbox/code/public/ldap3, inifile:
collected 1 item

test-encoding.py F                                                                                                                                                [100%]

=============================================================================== FAILURES ================================================================================
____________________________________________________________________________ test_encodings _____________________________________________________________________________

    def test_encodings():
        assert sys.getdefaultencoding() == 'ascii'
>       assert ldap3.utils.config._DEFAULT_CLIENT_ENCODING == 'UTF-8'
E       AssertionError: assert 'ascii' == 'UTF-8'
E         - ascii
E         + UTF-8

test-encoding.py:7: AssertionError
======================================================================= 1 failed in 0.13 seconds ========================================================================
$ python2.7 -m rwt -q pytest ldap3 -- -m pytest -s test-encoding.py
========================================================================== test session starts ==========================================================================
platform darwin -- Python 2.7.14, pytest-3.5.0, py-1.5.3, pluggy-0.6.0
rootdir: /Users/jaraco/Dropbox/code/public/ldap3, inifile:
collected 1 item

test-encoding.py .

======================================================================= 1 passed in 0.09 seconds ========================================================================
```

**Comment by cannatag** [comment]
In Windows I get cp850 as default encoding. It’s a Windows 10 machine with an Italian localization.

**Comment by jaraco** [comment]
I see. So maybe it would be easier to detect the issue if the offending character was ☃  instead of a valid character in your localization.

When I look at the [code in ava_to_dict](https://github.com/cannatag/ldap3/blob/aacdf9ec432d2705148b5d2ae9655fe77ab50164/ldap3/protocol/convert.py#L87), I notice that it's using implicit unicode/bytes conversion. It first calls `str(value)`, which converts the value to `bytes`. That's passed to escape_filter_chars, which converts the value to unicode using `convert.to_unicode`. If `sys.getdefaultencoding()` doesn't match `ldap3.get_config_param('DEFAULT_ENCODING')`, that's where an error will occur.

**Comment by jaraco** [comment]
We also have a production app that's experiencing this issue. It's a Python 2.7 app in an environment where `sys.stdin.encoding` is `ANSI_X3.4-1968`:

```
vuser@c64697bee75e:~$ /app/.heroku/venv/bin/python
Python 2.7.12 (default, Dec  4 2017, 14:50:18)
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.getdefaultencoding()
'ascii'
>>> sys.stdin.encoding
'ANSI_X3.4-1968'
>>> import ldap3.utils.config
>>> ldap3.utils.config._DEFAULT_CLIENT_ENCODING
'ANSI_X3.4-1968'
```

In this environment, unicode characters in the search or filter are encoded using UTF-8 but then attempted to be decoded using _DEFAULT_CLIENT_ENCODING, and only since ldap3 2.4.1 (2.4.0 works fine).

**Comment by jaraco** [comment]
This dockerfile will replicate the error:

```
from ubuntu:xenial

RUN apt update; apt upgrade -y
RUN apt install -y python2.7-dev wget
RUN wget https://bootstrap.pypa.io/get-pip.py -O - | python2.7
RUN python2.7 -m pip install ldap3==2.4.1

CMD python2.7 -c "import sys; print(sys.stdin.encoding); import ldap3; print(ldap3.utils.config._DEFAULT_CLIENT_ENCODING); conn=ldap3.Connection('ldap://ldap.forumsys.com', user='cn=read-only-admin,dc=example,dc=com', password='password'); conn.open(); print(conn.bind()); conn.search(u'OU=Root', u'(uid=\xfc)')"
```

Run with `docker build . -t foo; docker run -it foo`, it emits:

```
ANSI_X3.4-1968
ANSI_X3.4-1968
True
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/ldap3/core/connection.py", line 775, in search
    response = self.post_send_search(self.send('searchRequest', request, controls))
  File "/usr/local/lib/python2.7/dist-packages/ldap3/strategy/base.py", line 303, in send
    self.connection.request = BaseStrategy.decode_request(message_type, request, controls)
  File "/usr/local/lib/python2.7/dist-packages/ldap3/strategy/base.py", line 635, in decode_request
    result = search_request_to_dict(component)
  File "/usr/local/lib/python2.7/dist-packages/ldap3/operation/search.py", line 516, in search_request_to_dict
    'filter': filter_to_string(request['filter']),
  File "/usr/local/lib/python2.7/dist-packages/ldap3/operation/search.py", line 474, in filter_to_string
    ava = ava_to_dict(filter_object['equalityMatch'])
  File "/usr/local/lib/python2.7/dist-packages/ldap3/protocol/convert.py", line 87, in ava_to_dict
    return {'attribute': str(ava['attributeDesc']), 'value': escape_filter_chars(str(ava['assertionValue']))}
  File "/usr/local/lib/python2.7/dist-packages/ldap3/utils/conv.py", line 91, in escape_filter_chars
    text = to_unicode(text, encoding)
  File "/usr/local/lib/python2.7/dist-packages/ldap3/utils/conv.py", line 64, in to_unicode
    raise UnicodeError("Unable to convert client data to unicode: %r" % obj)
UnicodeError: Unable to convert client data to unicode: '\xc3\xbc'
```

**Comment by cannatag** [comment]
Hi, thanks for the dockerfile, I've replicated the error. Can you check the code in dev now? It should pass your docker test.

**Comment by jaraco** [comment]
Good news. That seems also to have corrected the issue in our test environment. I'm fairly confident that will also correct the issue in our production environment, though we won't know for sure until we roll out a release with this change. I'll ask our devs to do that and report back when we know more.

**Comment by cannatag** [comment]
I will release the fix in version 2.5. The problem affects Python 2 only. I should have identified the root cause, but I’m not sure if it needs more refinement.

**Comment by jaraco** [comment]
We've confirmed that this change also addresses the failure in our production environment. We're eager to see the release. We're currently blacklisting `ldap3==2.4.1`. If there will be any more releases not containing the fix, can you let us know so we can blacklist those as well?

**Comment by cannatag** [comment]
The fix should go in the upcoming 2.5 version. I should release it in the weekend.

**Comment by cannatag** [comment]
Fixed in 2.5. Sorry for delay.
